analyzers/src/SonarAnalyzer.CSharp/Rules/Utilities/TokenTypeAnalyzer.cs

```diff
@@ -102,17 +104,133 @@ token.Parent switch
+                {
+                    // some identifier can be bound to a type or a constant:
+                    CaseSwitchLabelSyntax => ClassifyIdentifierByModel(name), // case i:
+                    var x when ConstantPatternSyntaxWrapper.IsInstance(x) => ClassifyIdentifierByModel(name), // is i
+                    // nameof(i) can be bound to a type or a member
+                    ArgumentSyntax
+                    {
+                        Parent: ArgumentListSyntax
+                        {
+                            Parent: InvocationExpressionSyntax { Expression: IdentifierNameSyntax { Identifier.Text: "nameof" } }
+                        }
+                    } => ClassifyIdentifierByModel(name),
+                    // walk up memberaccess to detect cases like above
+                    MemberAccessExpressionSyntax x => ClassifySimpleNameExpressionSpecialContext(x, name),
+                    _ => ClassifySimpleNameExpressionSpecialNames(name)
+                };
+
+            /// <summary>
+            /// Some expression identifier are classified differently, like "value" in a setter.
+            /// </summary>
+            private TokenType ClassifySimpleNameExpressionSpecialNames(SimpleNameSyntax name) =>
+                name switch
+                {
+                    // "value" in a setter is a classified as keyword
+                    IdentifierNameSyntax { Identifier.Text: "value", Parent: not MemberAccessExpressionSyntax } when
+                        SemanticModel.GetSymbolInfo(name).Symbol is IParameterSymbol
+                        {
+                            ContainingSymbol: IMethodSymbol
+                            {
+                                MethodKind: MethodKind.PropertySet or MethodKind.EventAdd or MethodKind.EventRemove
+                            }
+                        } => TokenType.Keyword,
+                    _ => TokenType.UnknownTokentype,
+                };
+
+            private TokenType? ClassifyMemberAccess(SimpleNameSyntax name) =>
+                name switch
+                {
+                    // Most right hand side of a member access?
+                    { Parent: MemberAccessExpressionSyntax { Parent: not MemberAccessExpressionSyntax } x } when x.Name == name => ClassifySimpleNameExpressionSpecialContext(x, name),
+                    { } x => ClassifyIdentifierByModel(x)
+                };
+
+            private TokenType ClassifyIdentifierByModel(SimpleNameSyntax x) =>
+                SemanticModel.GetSymbolInfo(x).Symbol is INamedTypeSymbol
+                    ? TokenType.TypeName
+                    : TokenType.UnknownTokentype;
+
+            private static TokenType? ClassifySimpleNameType(SimpleNameSyntax x) =>
+                null;
+
+            private static bool IsInTypeContext(SimpleNameSyntax name) =>
+                // Based on Syntax.xml search for Type="TypeSyntax" and Type="NameSyntax"
+                name.Parent switch
```

Comment:
This is on the hotpath. Some imprvements here are possible, but not yet needed:
* Check for `Kind` instead of type for sealed classes
* Avoid the wrapper (IsInstance is allocating) by checking for `Kind` instead.
* Return `false` for the most commonly used expressions, in the beginning, to avoid the remaining checks for common cases (e.g. invocations).